### PR TITLE
Mask R-CNN on TPU

### PR DESCRIPTION
**Pull request by sprt**
This contains the work I've done so far to enable Mask R-CNN on TPU. I still have some more code to commit so I suggest leaving this open and reviewing the commits on a rolling basis, wdyt?

### CODE REVIEW

File: references/detection/train_tpu.py [hunk 1]
```diff
@@ -133,6 +133,7 @@ def do_prediction(image_path, use_cpu=False):
     model.eval()
 
     image = Image.open(image_path)
+    image = torchvision.transforms.functional.resize(image, (800, 600))
```
Comment (jysohn23) [inline references/detection/train_tpu.py:136]: FYI, once we get to training we'll need to do this exact thing in the input pipeline (but I believe it should already be there).

File: references/detection/train_tpu.py [hunk 2]
```diff
@@ -133,6 +133,7 @@ def do_prediction(image_path, use_cpu=False):
...
     image_tensor = torchvision.transforms.functional.to_tensor(image)
 
     if not use_cpu:
```
Comment (jysohn23) [inline references/detection/train_tpu.py:139]: Not needed for this PR, but we probably don't need `torch._XLAC._xla_set_default_device(...)` anymore as that's been updated.

File: torchvision/models/detection/roi_heads.py [hunk 3]
```diff
@@ -603,9 +603,7 @@ def postprocess_detections(self, class_logits, box_regression, proposals, image_
             boxes, scores, labels = boxes[keep], scores[keep], labels[keep]
 
             # non-maximum suppression, independently done per class
-            keep = box_ops.batched_nms(boxes, scores, labels, self.nms_thresh)
-            # keep only topk scoring predictions
-            keep = keep[:self.detections_per_img]
+            keep = box_ops.batched_nms(boxes, scores, labels, self.nms_thresh, self.detections_per_img)
```
Comment (jysohn23) [inline torchvision/models/detection/roi_heads.py:606]: Is `len(boxes) >= self.detections_per_img` pretty much guaranteed at training time? Wondering since we have added a check in the nms C++ implementation for it.

File: torchvision/models/detection/rpn.py [hunk 4]
```diff
@@ -60,7 +60,7 @@ def __init__(
         self.sizes = sizes
         self.aspect_ratios = aspect_ratios
         self.cell_anchors = None
-        self._cache = {}
+        # self._cache = {}
```
Comment (jysohn23) [inline torchvision/models/detection/rpn.py:63]: Even if we uncomment all the caching behavior, starting with the 3rd run (cache miss, cache hit, *cache hit*) we hit the compilation cache right? So performance would be fine?

File: torchvision/models/detection/rpn.py [hunk 5]
```diff
@@ -355,12 +355,9 @@ def filter_proposals(self, proposals, objectness, image_shapes, num_anchors_per_
         final_scores = []
         for boxes, scores, lvl, img_shape in zip(proposals, objectness, levels, image_shapes):
             boxes = box_ops.clip_boxes_to_image(boxes, img_shape)
-            keep = box_ops.remove_small_boxes(boxes, self.min_size)
-            boxes, scores, lvl = boxes[keep], scores[keep], lvl[keep]
             # non-maximum suppression, independently done per level
-            keep = box_ops.batched_nms(boxes, scores, lvl, self.nms_thresh)
+            keep = box_ops.batched_nms(boxes, scores, lvl, self.nms_thresh, self.post_nms_top_n)
```
Comment (jysohn23) [inline torchvision/models/detection/rpn.py:359]: Ditto.

**Comment by jysohn23** [comment]
Nice work 😄

File: references/detection/train_tpu.py (hunk 1 above)
Comment (sprt) [reply references/detection/train_tpu.py:136]: Sure thing, this is really for prototyping, long term I think this should go somewhere like [`GeneralizedRCNNTransform.resize()`](https://github.com/jysohn23/vision/blob/b45cdbf/torchvision/models/detection/transform.py#L55) where they already do some resizing based on min/max size thresholds.

File: torchvision/models/detection/rpn.py (hunk 4 above)
Comment (sprt) [reply torchvision/models/detection/rpn.py:63]: Yup exactly, I really just disabled this for now because it makes prototyping a little bit faster for me at the moment, as described in the commit message :)

File: references/detection/train_tpu.py (hunk 2 above)
Comment (sprt) [reply references/detection/train_tpu.py:139]: Hmm not sure what you're referring to here, I don't think we call that function anywhere?

Comment (jysohn23) [reply references/detection/train_tpu.py:139]: Ah sorry that was confusing. Check line 142.

File: torchvision/models/detection/rpn.py (hunk 5 above)
Comment (sprt) [reply torchvision/models/detection/rpn.py:359]: Basically you need to ensure that the hyperparameters you choose will enforce that `len(boxes) >= self.post_nms_top_n`. It's not that hard: if it's not the case, just make sure to generate more boxes by increasing the resize size, decreasing the stride, etc. Also if the assert is satisfied for one image, it'll be for all of them. See Finding 5 for more detail.

Comment (jysohn23) [reply torchvision/models/detection/rpn.py:359]: We can do it in a followup CL, but can we have a check to generate more boxes if `len(boxes) < self.post_nms_top_n`?

**Review by jysohn23 — APPROVED** [review]
